File: backend/Bussiness/Services/TourAndTravels/ItineraryBuilderService.cs
using System.Collections.Generic;
using Repository.Interfaces.TourAndTravels;
using static Domain.Models.TourAndTravels.ItineraryBuilder;

namespace Bussiness.Services.TourAndTravels
{
    public interface IItineraryBuilderService
    {
        BuilderResponse Save(SaveBuilderRequest request, long tenantId, long userId);
        BuilderResponse? Update(long id, SaveBuilderRequest request, long userId);
        BuilderResponse? GetById(long id);
        List<BuilderListItem> GetAll(long tenantId);
        bool Delete(long id);
        bool UpdateStatus(long id, string status);
        PricingSummary CalculatePricing(BuilderResponse itinerary, long tenantId);
    }

    public class ItineraryBuilderService : IItineraryBuilderService
    {
        private readonly IItineraryBuilderRepository _repo;
        private readonly ISeasonRepository _seasonRepo;
        private readonly ILocationRepository _locationRepo;

        public ItineraryBuilderService(
            IItineraryBuilderRepository repo,
            ISeasonRepository seasonRepo,
            ILocationRepository locationRepo)
        {
            _repo = repo;
            _seasonRepo = seasonRepo;
            _locationRepo = locationRepo;
        }

        public BuilderResponse Save(SaveBuilderRequest request, long tenantId, long userId)
            => _repo.Save(request, tenantId, userId);

        public BuilderResponse? Update(long id, SaveBuilderRequest request, long userId)
            => _repo.Update(id, request, userId);

        public BuilderResponse? GetById(long id)
            => _repo.GetById(id);

        public List<BuilderListItem> GetAll(long tenantId)
            => _repo.GetAll(tenantId);

        public bool Delete(long id) => _repo.Delete(id);

        public bool UpdateStatus(long id, string status)
            => _repo.UpdateStatus(id, status);

        public PricingSummary CalculatePricing(BuilderResponse itinerary, long tenantId)
        {
            decimal accommodation = 0, transport = 0, activities = 0, meals = 0, guide = 0, others = 0;

            foreach (var day in itinerary.Days)
            {
                foreach (var item in day.Items)
                {
                    decimal cost = item.UnitPrice * item.Quantity;
                    switch (item.ItemType.ToLower())
                    {
                        case "hotel": accommodation += cost; break;
                        case "transport": transport += cost; break;
                        case "activity": activities += cost; break;
                        case "meal": meals += cost; break;
                        case "guide": guide += cost; break;
                        default: others += cost; break;
                    }
                }
            }

            // Season multiplier
            decimal seasonMultiplier = 1.0m;
            string? seasonName = null;
            if (itinerary.SeasonId.HasValue)
            {
                var season = _seasonRepo.GetById(itinerary.SeasonId.Value);
                if (season != null)
                {
                    seasonMultiplier = season.PriceMultiplier;
                    seasonName = $"{season.Name} ({season.PriceMultiplier:0.00}x)";
                }
            }

            // Location multiplier (average across days)
            decimal locationMultiplier = 1.0m;
            var locationIds = itinerary.Days
                .Where(d => d.LocationId.HasValue)
                .Select(d => d.LocationId!.Value)
                .Distinct()
                .ToList();

            if (locationIds.Any())
            {
                var locations = _locationRepo.GetAll(tenantId)
                    .Where(l => locationIds.Contains(l.Id))
                    .ToList();

                if (locations.Any())
                    locationMultiplier = locations.Average(l => l.CostMultiplier);
            }

            decimal totalMultiplier = seasonMultiplier * locationMultiplier;
            decimal subtotal = (accommodation + transport + activities + meals + guide + others) * totalMultiplier;
            decimal markupAmount = subtotal * itinerary.MarkupPercentage / 100m;
            decimal total = subtotal + markupAmount;
            int numPax = itinerary.NumPax > 0 ? itinerary.NumPax : 2;

            return new PricingSummary
            {
                Accommodation = accommodation,
                Transport = transport,
                Activities = activities,
                Meals = meals,
                Guide = guide,
                Others = others,
                Subtotal = Math.Round(subtotal, 2),
                MarkupPercentage = itinerary.MarkupPercentage,
                MarkupAmount = Math.Round(markupAmount, 2),
                TotalEstimatedPrice = Math.Round(total, 2),
                PricePerPerson = Math.Round(total / numPax, 2),
                NumPax = numPax,
                SeasonName = seasonName,
                SeasonMultiplier = seasonMultiplier,
                LocationMultiplier = Math.Round(locationMultiplier, 2),
                TotalMultiplier = Math.Round(totalMultiplier, 2)
            };
        }
    }
}

File: frontend/src/app/pages/itinerary-builder/itinerary-builder.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

// ═══════════════════════════════════════════
// BUILDER INTERFACES
// ═══════════════════════════════════════════
export interface BuilderItem {
    id?: number;
    itemType: string;        // hotel, transport, activity, meal, guide
    inventoryItemId?: number;
    itemName: string;
    itemDetails?: string;
    sortOrder: number;
    quantity: number;
    unitPrice: number;
    currency: string;
    notes?: string;
    /** For hotel items: how many pax the unit price covers (used to calculate rooms needed) */
    paxCount?: number;
}

export interface BuilderDay {
    id?: number;
    dayNumber: number;
    title?: string;
    location?: string;
    locationId?: number;
    accommodationCategoryId?: number;
    accommodationCategoryName?: string;
    notes?: string;
    items: BuilderItem[];
    expanded: boolean;       // UI-only
}

export interface SaveBuilderRequest {
    title: string;
    description?: string;
    durationDays: number;
    defaultCurrency: string;
    seasonId?: number;
    markupPercentage: number;
    numPax: number;
    status: string;
    difficultyLevel?: string;
    days: {
        dayNumber: number;
        title?: string;
        location?: string;
        locationId?: number;
        accommodationCategoryId?: number;
        notes?: string;
        items: {
            itemType: string;
            inventoryItemId?: number;
            itemName: string;
            itemDetails?: string;
            sortOrder: number;
            quantity: number;
            unitPrice: number;
            currency: string;
            notes?: string;
            /** For hotel items: how many pax the unit price covers (used to calculate rooms needed) */
            paxCount?: number;
        }[];
    }[];
}

export interface BuilderResponse {
    id: number;
    title: string;
    description?: string;
    durationDays: number;
    defaultCurrency: string;
    seasonId?: number;
    markupPercentage: number;
    numPax: number;
    status: string;
    difficultyLevel?: string;
    isActive: boolean;
    createdAt: string;
    days: {
        id: number;
        dayNumber: number;
        title?: string;
        location?: string;
        locationId?: number;
        accommodationCategoryId?: number;
        accommodationCategoryName?: string;
        notes?: string;
        items: {
            id: number;
            itemType: string;
            inventoryItemId?: number;
            itemName: string;
            itemDetails?: string;
            sortOrder: number;
            quantity: number;
            unitPrice: number;
            currency: string;
            notes?: string;
            /** For hotel items: how many pax the unit price covers (used to calculate rooms needed) */
            paxCount?: number;
        }[];
    }[];
}

export interface BuilderListItem {
    id: number;
    title: string;
    durationDays: number;
    status: string;
    defaultCurrency: string;
    numPax: number;
    isActive: boolean;
    createdAt: string;
}

export interface PricingSummary {
    accommodation: number;
    transport: number;
    activities: number;
    meals: number;
    guide: number;
    others: number;
    subtotal: number;
    markupPercentage: number;
    markupAmount: number;
    totalEstimatedPrice: number;
    pricePerPerson: number;
    numPax: number;
    seasonName?: string;
    seasonMultiplier: number;
    locationMultiplier: number;
    totalMultiplier: number;
}

// ═══════════════════════════════════════════
// INVENTORY LIBRARY ITEMS (loaded from existing APIs)
// ═══════════════════════════════════════════
export interface InventoryHotel {
    id: number;
    name: string;
    location: string;
    starRating?: number;
    category?: string;
    rooms?: { roomType: string; pricePerNight: number; capacity: number }[];
}

export interface InventoryVehicle {
    id: number;
    vehicleType: string;
    model: string;
    capacity: number;
    pricePerDay: number;
}

export interface InventoryActivity {
    id: number;
    name: string;
    activityType: string;
    location?: string;
    pricePerPerson: number;
    durationHours: number;
}

export interface InventoryGuide {
    id: number;
    fullName: string;
    phone?: string;
    email?: string;
    specialization?: string;
    languages?: string[];
    pricePerDay: number;
    rating?: number;
    experienceYears?: number;
    bio?: string;
}

export interface LocationItem {
    id: number;
    name: string;
    country: string;
    costMultiplier: number;
}

export interface SeasonItem {
    id: number;
    name: string;
    seasonType: string;
    startDate: string;
    endDate: string;
    priceMultiplier: number;
}

export interface CurrencyItem {
    id: number;
    code: string;
    name: string;
    symbol: string;
}

export interface AccommodationCategory {
    id: number;
    name: string;
    code: string;
    icon?: string;
    starRating?: number;
    sortOrder: number;
    description?: string;
    isActive: boolean;
    avgPricePerNight?: number;
    baseCurrency?: string;
    /** Number of pax the avg price covers (e.g. 2 = price per room for 2 people) */
    paxCount: number;
}

export interface HotelSuggestion {
    hotelId: number;
    hotelName: string;
    location?: string;
    starRating?: number;
    category?: string;
    accommodationCategoryId?: number;
    accommodationCategoryName?: string;
    minPrice?: number;
    description?: string;
}

// ═══════════════════════════════════════════
// SERVICE
// ═══════════════════════════════════════════
@Injectable({ providedIn: 'root' })
export class ItineraryBuilderService {
    private api = `${environment.apiBaseUrl}/itinerary-builder`;
    private invApi = `${environment.apiBaseUrl}`;

    constructor(private http: HttpClient) {}

    // Builder CRUD
    save(req: SaveBuilderRequest): Observable<BuilderResponse> {
        return this.http.post<BuilderResponse>(this.api, req);
    }

    update(id: number, req: SaveBuilderRequest): Observable<BuilderResponse> {
        return this.http.put<BuilderResponse>(`${this.api}/${id}`, req);
    }

    getById(id: number): Observable<BuilderResponse> {
        return this.http.get<BuilderResponse>(`${this.api}/${id}`);
    }

    getAll(): Observable<BuilderListItem[]> {
        return this.http.get<BuilderListItem[]>(this.api);
    }

    delete(id: number): Observable<void> {
        return this.http.delete<void>(`${this.api}/${id}`);
    }

    calculatePricing(id: number): Observable<PricingSummary> {
        return this.http.post<PricingSummary>(`${this.api}/${id}/pricing`, {});
    }

    pricingPreview(data: any): Observable<PricingSummary> {
        return this.http.post<PricingSummary>(`${this.api}/pricing-preview`, data);
    }

    // Inventory library (existing APIs)
    getHotels(): Observable<InventoryHotel[]> {
        return this.http.get<InventoryHotel[]>(`${this.invApi}/inventory/hotels`);
    }

    getVehicles(): Observable<InventoryVehicle[]> {
        return this.http.get<InventoryVehicle[]>(`${this.invApi}/inventory/vehicles`);
    }

    getActivities(): Observable<InventoryActivity[]> {
        return this.http.get<InventoryActivity[]>(`${this.invApi}/inventory/activities`);
    }

    getGuides(): Observable<InventoryGuide[]> {
        return this.http.get<InventoryGuide[]>(`${this.invApi}/inventory/guides`);
    }

    getLocations(): Observable<LocationItem[]> {
        return this.http.get<LocationItem[]>(`${this.invApi}/inventory/locations`);
    }

    getSeasons(): Observable<SeasonItem[]> {
        return this.http.get<SeasonItem[]>(`${this.invApi}/inventory/seasons`);
    }

    getCurrencies(): Observable<CurrencyItem[]> {
        return this.http.get<CurrencyItem[]>(`${this.invApi}/inventory/currencies`);
    }

    getAccommodationCategories(): Observable<AccommodationCategory[]> {
        return this.http.get<AccommodationCategory[]>(`${this.invApi}/inventory/accommodation-categories`);
    }

    getHotelsByCategory(categoryId: number, locationId?: number): Observable<HotelSuggestion[]> {
        let url = `${this.invApi}/inventory/accommodation-categories/${categoryId}/hotels`;
        if (locationId) url += `?locationId=${locationId}`;
        return this.http.get<HotelSuggestion[]>(url);
    }
}
